File: minHeap.c
#include <stdbool.h>
#include "utils.h"

/*
values = [1,2,3,4,5,6,7]
indices= [0,1,2,3,4,5,6]

      1
    /   \
   2     3
  / \   / \
 4   5 6   7
*/

typedef struct {
  int *elements;
  int size;
} Heap;

int parent(int index) {
  return (index - 1) / 2;
}

int leftChild(int index) {
  return (2 * index) + 1;
}

int rightChild(int index) {
  return (2 * index) + 2;
}

bool insert(Heap *heap, int value) {
  return true;
}

bool extractMin(Heap *heap, int *value) {
  *value = heap->elements[0];

  int lastValue = heap->elements[heap->size - 1];


  return true;
}

bool bubbleDown(int *elements) {
  
}

bool heapify(Heap *heap) {
  for (int i = 0; i < heap->size; i++) {

  }

  return true;
}

bool buildHeap(int *elements, int size) {
  Heap *heap = (Heap *)safeMalloc(sizeof(Heap));
  int *heapElements = safeCalloc(sizeof(int), size);

  heap->elements = heapElements;
  heap->size = size;

  for (int i = 0; i < size; i++) {
    heapElements[i] = elements[i];
  }

  heapify(heap);

  return true;
}

File: minHeap.py
class MinHeap:
    def __init__(self, elements):
        self.elements = self.build(elements)
        self.size = 0

    def insert():
        return True

    def delete():
        return True

    def heapify(self, elements, size, index):
        # If left child is less than, swap them and heapify left child
        if self.leftChild(index) < elements[]


        # If right child is less than, swap them and heapify right child
        return True

    def parent(self, index):
        idx = (index / 2) - 1
        return self.elements[idx]

    def leftChild(self, index):
        idx = (index * 2) + 1
        return self.elements[idx]

    def rightChild(self, index):
        idx = (index * 2) + 2
        return self.elements[idx]

    def build(self, elements, size):
        for i in range(self.parent(size), 0, -1):
            self.heapify(elements, size, i)
